internal/provider/resource_hvn_route.go

```diff
@@ -1,14 +1,293 @@
+	}
+	log.Printf("Target %s is found, proceeding with creating HVN route for HVN (%s).", target, hvnLink.ID)
+
+	targetLink.Location.Region = retrievedHvn.Location.Region
+
+	hvnRouteParams := network_service.NewCreateHVNRouteParams()
+	hvnRouteParams.Context = ctx
+	hvnRouteParams.HvnLocationOrganizationID = loc.OrganizationID
+	hvnRouteParams.HvnLocationProjectID = loc.ProjectID
+	hvnRouteParams.HvnID = hvnLink.ID
+	hvnRouteParams.Body = &networkmodels.HashicorpCloudNetwork20200907CreateHVNRouteRequest{
+		Destination: destination,
+		Hvn:         hvnLink,
+		ID:          hvnRouteID,
+		Target: &networkmodels.HashicorpCloudNetwork20200907HVNRouteTarget{
+			HvnConnection: targetLink,
+		},
+	}
+	log.Printf("[INFO] Creating HVN route for HVN (%s) with destination CIDR %s", hvnLink.ID, destination)
+	hvnRouteResp, err := client.Network.CreateHVNRoute(hvnRouteParams, nil)
+	if err != nil {
+		return diag.Errorf("unable to create HVN route for HVN (%s) with destination CIDR %s: %v", hvnLink.ID, destination, err)
+	}
+
+	hvnRoute := hvnRouteResp.Payload.Route
+
+	// Set the globally unique id of this HVN route in the state now since it has
+	// been created, and from this point forward should be deletable.
+	link := newLink(hvnRoute.Hvn.Location, HVNRouteResourceType, hvnRoute.ID)
+	url, err := linkURL(link)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(url)
+
+	// Wait for HVN route to be created.
+	if err := clients.WaitForOperation(ctx, client, "create HVN route", loc, hvnRouteResp.Payload.Operation.ID); err != nil {
+		return diag.Errorf("unable to create HVN route for HVN (%s) with destination CIDR %s:%v", hvnLink.ID, destination, err)
+	}
+
+	log.Printf("[INFO] Created HVN route for HVN (%s) with destination CIDR %s", hvnRoute.ID, hvnRoute.Hvn.ID)
+
+	hvnRoute, err = clients.WaitForHVNRouteToBeActive(ctx, client, destination, hvnLink.ID, loc, d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := setHVNRouteResourceData(d, hvnRoute, loc); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func resourceHvnRouteRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client)
+
+	hvn := d.Get("hvn").(string)
+	var hvnLink *sharedmodels.HashicorpCloudLocationLink
+
+	hvnLink, err := parseLinkURL(hvn, HvnResourceType)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	destination := d.Get("destination_cidr").(string)
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      client.Config.ProjectID,
+	}
+
+	route, err := clients.ListHVNRoutes(ctx, client, hvnLink.ID, destination, "", "", loc)
+	if err != nil {
+		if clients.IsResponseCodeNotFound(err) {
+			log.Printf("[WARN] HVN route for HVN (%s) with destination CIDR %s not found, removing from state", hvnLink.ID, destination)
```

Comment:
@bcmdarroch is removing datasource from the state is an idiomatic thing to do in this case?

Reply:
should we just return an error and fail tf apply? otherwise we just continue operations hoping that user would read output

Reply:
Yes, this is a pattern. If it the API does not return the route, that means it was deleted and thus TF needs to update its state to reflect that. Setting the ID to nil in the line below is the same as deleting it from the state. The error in L238 should then be displayed, so the user will know there was a change in the state.